identity-iota/src/credential/deficiencies.rs

```diff
@@ -0,0 +1,188 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use num_traits::FromPrimitive;
+use serde::Serialize;
+use std::fmt::Display;
+
+#[derive(PartialEq, Eq, Hash, Clone, Copy, Debug, Serialize, num_derive::FromPrimitive)]
+#[non_exhaustive]
+/// A possibly acceptable deficiency a credential might have.
+pub enum CredentialDeficiency {
+  #[doc(hidden)] // hidden until we have decided how to act on deactivated documents
+  /// At least one subject document is deactivated
+  DeactivatedSubjectDocuments = 0,
+  /// The credential has expired
+  Expired = 1,
+  /// The credential has not yet become active
+  // (issuance_date is in the future)
+  Dormant = 2,
+}
+
+impl CredentialDeficiency {
+  /// Provides a description of the category
+  pub fn description(&self) -> &'static str {
+    match *self {
+      Self::DeactivatedSubjectDocuments => "contains subjects with deactivated DID documents",
+      Self::Expired => "the expiry date is in the past",
+      Self::Dormant => "the activation date is in the future",
+    }
+  }
+  // The number of refutation categories. We do not use strum for this as we do not want to unnecessarily pollute the
+  // public API
+  pub(super) const COUNT: usize = 3;
+}
+
+impl Display for CredentialDeficiency {
+  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+    write!(f, "{}", self.description())
+  }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
+/// A set containing zero or more variants of [`CredentialDeficiency`]
+pub struct CredentialDeficiencySet {
```

Comment:
Discussion: I'm not sure about reimplementing collections. I guess this is a bit more ergonomic than taking `&[CredentialDeficiency]` everywhere? But that seems fine to me, and wouldn't require implementing collection-specific things. Perhaps I'm missing some usecase? Not a strong opinion overall.

Reply:
Good points. You have made me realise that `CredentialDeficiencySet` needs rethinking.